Entities/Common/Emotes/EmoteBubble.as

```diff
@@ -7,61 +7,85 @@ void onInit(CBlob@ blob)
 	blob.addCommandID("emote");
 
 	CSprite@ sprite = blob.getSprite();
-	blob.set_u8("emote", Emotes::off);
+	blob.set_string("emote", "");
 	blob.set_u32("emotetime", 0);
-	//init emote layer
-	CSpriteLayer@ emote = sprite.addSpriteLayer("bubble", "Entities/Common/Emotes/Emoticons.png", 32, 32, 0, 0);
-	emote.SetIgnoreParentFacing(true);
-	emote.SetFacingLeft(false);
 
-	if (emote !is null)
+	dictionary packs;
+	if (!getRules().get("emote packs", packs)) return;
+	string[] tokens = packs.getKeys();
+
+	for (uint i = 0; i < tokens.size(); i++)
 	{
-		emote.SetOffset(Vec2f(0, -sprite.getBlob().getRadius() * 1.5f - 16));
-		emote.SetRelativeZ(100.0f);
-		{
-			Animation@ anim = emote.addAnimation("default", 0, true);
+		EmotePack@ pack;
+		packs.get(tokens[i], @pack);
+
+		//init emote layer
+		CSpriteLayer@ layer = sprite.addSpriteLayer("bubble" + pack.token, pack.filePath, 32, 32, 0, 0);
+		layer.SetIgnoreParentFacing(true);
+		layer.SetFacingLeft(false);
 
-			for (int i = 0; i < Emotes::emotes_total; i++)
+		if (layer !is null)
+		{
+			layer.SetOffset(Vec2f(0, -sprite.getBlob().getRadius() * 1.5f - 16));
+			layer.SetRelativeZ(100.0f);
 			{
-				anim.AddFrame(i);
+				Animation@ anim = layer.addAnimation("default", 0, true);
+
+				for (int i = 0; i < pack.emotes.size(); i++)
+				{
+					anim.AddFrame(i);
+				}
 			}
+			layer.SetVisible(false);
+			layer.SetHUD(true);
 		}
-		emote.SetVisible(false);
-		emote.SetHUD(true);
 	}
 }
 
 void onTick(CBlob@ blob)
 {
 	blob.getCurrentScript().tickFrequency = 6;
-	// if (blob.exists("emote"))	 will show skull if none existant
+
 	if (!blob.getShape().isStatic())
 	{
-		CSprite@ sprite = blob.getSprite();
-		CSpriteLayer@ emote = sprite.getSpriteLayer("bubble");
+		dictionary packs;
+		if (!getRules().get("emote packs", packs)) return;
+		string[] tokens = packs.getKeys();
 
-		const u8 index = blob.get_u8("emote");
-		if (is_emote(blob, index) && !blob.hasTag("dead") && !blob.isInInventory())
+		Emote@ emote = getEmote(blob.get_string("emote"));
+
+		for (uint i = 0; i < tokens.size(); i++)
```

Comment:
From what I understand, this is not needed. Emote already contains pack.token, so you can grab the bubble sprite layer without having to loop through all the packs.

Reply:
The reason I loop through all the packs is to hide the ones that aren't the correct pack.